From 6b227e932f9d42a4174be3c6564ef0457a32475f Mon Sep 17 00:00:00 2001
From: Sai Vemuri <svemuri@redhat.com>
Date: Thu, 11 Sep 2014 22:43:23 -0400
Subject: [infiniband] cxgb4: Fix memory leaks in c4iw_alloc() error paths

Message-id: <1410475447-94392-65-git-send-email-svemuri@redhat.com>
Patchwork-id: 93827
O-Subject: [RHEL7.1 PATCH BZ 1124947 064/108] RDMA/cxgb4: Fix memory leaks in c4iw_alloc() error paths
Bugzilla: 1124947
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

c4iw_alloc() bails out without freeing the storage that 'devp' points to.

Picked up by Coverity - CID 1204241.

Fixes: fa658a98a2 ("RDMA/cxgb4: Use the BAR2/WC path for kernel QPs and T5 devices")
Signed-off-by: Christoph Jaeger <christophjaeger@linux.com>
Acked-by: Steve Wise <swise@opengridcomputing.com>
Signed-off-by: Roland Dreier <roland@purestorage.com>
(cherry picked from commit 65b302ad31b02b0790417f4e65833af494cb35ce)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/hw/cxgb4/device.c b/drivers/infiniband/hw/cxgb4/device.c
index f4fa50a..8914ea9 100644
--- a/drivers/infiniband/hw/cxgb4/device.c
+++ b/drivers/infiniband/hw/cxgb4/device.c
@@ -736,6 +736,7 @@ static struct c4iw_dev *c4iw_alloc(const struct cxgb4_lld_info *infop)
    pci_resource_len(devp->rdev.lldi.pdev, 2));
   if (!devp->rdev.bar2_kva) {
    pr_err(MOD "Unable to ioremap BAR2\n");
+   ib_dealloc_device(&devp->ibdev);
    return ERR_PTR(-EINVAL);
   }
  } else if (ocqp_supported(infop)) {
@@ -747,6 +748,7 @@ static struct c4iw_dev *c4iw_alloc(const struct cxgb4_lld_info *infop)
    devp->rdev.lldi.vr->ocq.size);
   if (!devp->rdev.oc_mw_kva) {
    pr_err(MOD "Unable to ioremap onchip mem\n");
+   ib_dealloc_device(&devp->ibdev);
    return ERR_PTR(-EINVAL);
   }
  }
-- 
1.7.1